### DISCUSSION TITLE
Reload the current database?

### DISCUSSION THREAD

**Discussion by CaledoniaProject** [github/codeql#8563] [General]
The database is not reloaded when I run `codeql database create` and create the database in the same directory. I had to remove and open the database again, is this expected behavior?

Is it possible to do a reload instead of "remove and add"? Looks like there's no such context menu:
![screenshot 2022-03-26 at 08 56 54](https://user-images.githubusercontent.com/1357701/160218291-adf8d41e-5a63-4eee-9eaf-b85cd46661e9.jpg)

**Comment by smowton** [comment]
There isn't such an option yet. If you're in an edit-code, create-database, query-database cycle I find it's best to create db1, db2, ... so each db change is loading a new one, not removing and re-adding an old one.

**Reply by CaledoniaProject** [reply]
Are you planned to add one? It is a quite common scenario to reload the database when I recompile the source code. I prefer a context menu to do that.

**Comment by hmakholm** [comment]
We have some ongoing discussions of how to best improve this user experience.

A "reload" command does sound like an obvious thing to do from the outside, but there are some technical problem with making it work. Of course it could work at least as well as "close the database and open it again", but that itself is not guaranteed to work well.

The workflow we officially support is that you're supposed to close the database in VSC _before_ you delete and overwrite it. As long as the database is "open", the underlying query engine might be in a state where it thinks it needs to clean up after itself in a database directory at close. Then if that directory has been deleted and replaced with something else, the intended "clean-up" might instead end up breaking something in the _new_ database. This doesn't happen often, but the risk is there.

We are looking at a way to detect the sudden disappearance of the database underneath the engine, but it seems to be difficult to do it reliably without very platform-dependent hacks. As long as that hasn't been solved, we're loath to create an UI option that makes it look like it is okay to wait with notifying VCS until you _have_ replaced the database. (And once we find a solution, the UI option might not be needed after all -- one might hope that if we can _detect_ the database disappearing, we could react by _transparently_ reopening it).
